docs/polkadot/implementors-guide/guide.md

```diff
@@ -989,6 +989,67 @@ Dispatch a `PovFetchSubsystemMessage(relay_parent, candidate_hash, sender)` and
 
 (TODO: send statements to Statement Distribution subsystem, handle shutdown signal from candidate backing subsystem)
 
+### Statement Gossip Subsystem
+
+The statement gossip subsystem is responsible for gossiping out and receiving [Signed Statements](#Signed-statement-type) from validators.
+
+This subsystem communicates with the overseer via [two message types](#Candidate-Backing-Subsystem-Messages), one for overseer -> subsystem and another for subsystem -> overseer. Currently this just differentiates between statements that the subsystem has been instructed to gossip out, and statements that the subsystem has received.
+
+Here is one design for how it might look in code:
+
+```rust
+struct StatementGossipSubsystem<GS: GossipService, SP: Spawn> {
+    /// This trait comes from the legacy networking code, so it is likely to be changed at some point.
+    gossip_service: GS,
```

Comment:
cc @mxinden I'd expect that we'd do things the way they're laid out here, where the subsystem owns its entire gossip service.

However with the gossip-service as-is we need to write a `GossipValidator` that handles incoming messages and checks signatures and controls what messages we send to peers. Since the validator set can change from block to block, we'll want to:
  - only accept messages related to the relay-parents of jobs that we have active
  - fetch the validator set for a parent-hash as part of the job-starting process - either by having overseer provide this in `StartWork` or by calling out to another `RuntimeApi` subsystem to avoid shared ownership of the runtime API.
  - communicate to our peers which messages we are willing to receive and only send messages to peers they are expecting to receive. this is done with "neighbor packets" in the existing gossip code, which are special messages used to establish data-dependency, as @infinity0 frames it.

Reply:
I would imagine the `StatementGossipSubsystem` to own a [`GossipEngine`](https://github.com/paritytech/substrate/blob/master/client/network-gossip/src/bridge.rs#L37) as well as some means of communication to a custom implementation of the [`Validator`](https://github.com/paritytech/substrate/blob/master/client/network-gossip/src/validator.rs#L23) trait. I don't have an opinion whether the `GossipEngine` should be wrapped in a `GossipService`.

> - communicate to our peers which messages we are willing to receive and only send messages to peers they are expecting to receive. this is done with "neighbor packets" in the existing gossip code, which are special messages used to establish data-dependency, as Ximin frames it.

Sounds good to me. Only thing that would be important to me is that we don't misuse the notion of a `Validator` to send out messages like we do in `substrate/client/finality-grandpa`.

Reply:
As I understand the gossip model, it's the responsibility of the validator to re-broadcast messages that have been received. Without doing that, messages will not propagate. It's necessary to do that at the gossip-validator level so we can distinguish between neighbor packets and gossip messages. So without changing the lower-level gossip APIs substantially, we have to send messages from the gossip-validator. Or did you mean something different?

Reply:
> As I understand the gossip model, it's the responsibility of the validator to re-broadcast messages that have been received. Without doing that, messages will not propagate. 

Yes, even though I would like to be a bit nit picky and rephrase it as: "A validator validates gossip messages and decides whether a message (if and only if valid) should be further propagated (via the [`ValidationResult`](https://github.com/paritytech/substrate/blob/master/client/network-gossip/src/validator.rs#L76))."

A gossip validator only reacts to messages but is never proactive as in sending new messages (not like what we do today with new neighbor messages in grandpa). Sending new messages (e.g. neighbor messages or statements) should happen through the `GossipEngine`.